File: VB/ServerSide/Controllers/ReportDesignerController.vb
Imports System.Collections.Generic
Imports DevExpress.DataAccess.Sql
Imports DevExpress.Web.Mvc.Controllers
Imports DevExpress.XtraReports.Web.ReportDesigner
Imports System.Web.Mvc

Namespace ServerSide.Controllers
    Public Class ReportDesignerController
        Inherits ReportDesignerApiController

        ' This action processes each internal request from the Report Designer.
        Public Overrides Function Invoke() As ActionResult
            Dim result = MyBase.Invoke()
            ' Allow cross-domain requests.
            Response.AppendHeader("Access-Control-Allow-Origin", "*")
            Return result
        End Function

        ' Create an action to provide the Report Designer with the following initialization data:
        '  - A report (required)
        '  - Available data sources (optional)
        '  - URLs for Invoke actions of reporting controllers (required)
        Public Function GetReportDesignerModel(ByVal reportUrl As String) As ActionResult
            Response.AppendHeader("Access-Control-Allow-Origin", "*")

            Dim modelJsonScript As String = (New ReportDesignerClientSideModelGenerator()).GetJsonModelScript(reportUrl, GetAvailableDataSources(), "ReportDesigner/Invoke", "WebDocumentViewer/Invoke", "QueryBuilder/Invoke") ' The URI path of the controller action that processes requests from the Query Builder. -  The URI path of the controller action that processes requests from the Web Document Viewer. -  The URI path of the controller action that processes requests from the Report Designer. -  Obtain data sources that will be available in the Report Designer for adding to reports. -  Obtain a report to open in the Report Designer on application start.
            Return Content(modelJsonScript, "application/json")
        End Function

        Private Function GetAvailableDataSources() As Dictionary(Of String, Object)
            Dim dataSources = New Dictionary(Of String, Object)()
            Dim ds As New SqlDataSource("Northwind_Connection")
            Dim query = SelectQueryFluentBuilder.AddTable("Products").SelectAllColumns().Build("Products")
            ds.Queries.Add(query)
            ds.RebuildResultSchema()
            dataSources.Add("SqlDataSource", ds)
            Return dataSources
        End Function
    End Class
End Namespace

File: CS/ServerSide/Controllers/ReportDesignerController.cs
using System.Collections.Generic;
using System.Web.Mvc;
using DevExpress.DataAccess.Sql;
using DevExpress.Web.Mvc.Controllers;
using DevExpress.XtraReports.Web.ReportDesigner;

namespace ServerSide.Controllers {
    public class ReportDesignerController : ReportDesignerApiControllerBase {
        // This action processes each internal request from the Report Designer.
        public override ActionResult Invoke() {
            var result = base.Invoke();
            // Allow cross-domain requests.
            Response.AppendHeader("Access-Control-Allow-Origin", "*");
            return result;
        }

        // Create an action to provide the Report Designer with the following initialization data:
        //  - A report (required)
        //  - Available data sources (optional)
        //  - URLs for Invoke actions of reporting controllers (required)
        public ActionResult GetReportDesignerModel(string reportUrl) {
            Response.AppendHeader("Access-Control-Allow-Origin", "*");

            string modelJsonScript =
                new ReportDesignerClientSideModelGenerator()
                .GetJsonModelScript(
                    reportUrl,                 // Obtain a report to open in the Report Designer on application start.
                    GetAvailableDataSources(), // Obtain data sources that will be available in the Report Designer for adding to reports.
                    "ReportDesigner/Invoke",   // The URI path of the controller action that processes requests from the Report Designer.
                    "WebDocumentViewer/Invoke",// The URI path of the controller action that processes requests from the Web Document Viewer.
                    "QueryBuilder/Invoke"      // The URI path of the controller action that processes requests from the Query Builder.
                );
            return Content(modelJsonScript, "application/json");
        }

        Dictionary<string, object> GetAvailableDataSources() {
            var dataSources = new Dictionary<string, object>();
            SqlDataSource ds = new SqlDataSource("Northwind_Connection");
            var query = SelectQueryFluentBuilder.AddTable("Products").SelectAllColumns().Build("Products");
            ds.Queries.Add(query);
            ds.RebuildResultSchema();
            dataSources.Add("SqlDataSource", ds);
            return dataSources;
        }
    }
}